### PR TITLE
Support key authentication in ssh-transport 

### PR DESCRIPTION
**Pull request by petervo**
This is the start of the work to support key authentication. This adds support for for proxing an ssh-agent from the users session as seen by the localhost bridge to new ssh based transports opened by cockpit-ws.

This PR will not run or compile as it requires changes to libssh that i'm working to get in. I'm putting it here so @stefwalter and all can review the general approach and the code. I'm also working on tests.

### CODE REVIEW

File: src/bridge/bridge.c [hunk 1]
```diff
@@ -510,6 +519,7 @@ run_bridge (const gchar *interactive)
     g_object_unref (super);
   g_object_unref (pcp);
   g_object_unref (transport);
+  g_object_unref (auth_address);
```
Comment (stefwalter) [inline src/bridge/bridge.c]: This dereferences NULL.

File: src/ws/cockpitsshtransport.c [hunk 2]
```diff
@@ -513,6 +548,241 @@ cockpit_ssh_data_free (CockpitSshData *data)
   g_free (data);
 }
 
+
+/* ----------------------------------------------------------------------------
```
Comment (stefwalter) [inline src/ws/cockpitsshtransport.c]: Is there a reason this code can't live in a separate file? This file is pretty big already. If it's too hard, that's fine, your call.

File: src/ws/cockpitwebservice.c [hunk 3]
```diff
@@ -1017,7 +1017,7 @@ on_session_recv (CockpitTransport *transport,
   if (session == NULL)
     {
       /* This is not an error since channel closing can race */
-      g_debug ("dropping message with unknown channel %s from session", channel);
+      g_warning ("dropping message with unknown channel %s from session", channel);
```
Comment (stefwalter) [inline src/ws/cockpitwebservice.c]: What is the reason for changing this? Also the comment no longer matches the code.

File: src/bridge/bridge.c [hunk 4]
```diff
@@ -485,6 +487,13 @@ run_bridge (const gchar *interactive)
       super = cockpit_portal_new_superuser (transport);
     }
 
+  ssh_auth_sock = g_getenv ("SSH_AUTH_SOCK");
+  if (ssh_auth_sock != NULL && ssh_auth_sock[0] != '\0');
+    {
+      auth_address = g_unix_socket_address_new (ssh_auth_sock);
+      cockpit_channel_internal_address ("ssh-agent", auth_address);
```
Comment (stefwalter) [inline src/bridge/bridge.c]: We should start documenting these. What about a new section in ```doc/protocol.md```?

**Comment by stefwalter** [comment]
I like the idea of using an internal address. It's a good balance of pragmatic and useful.

**Comment by petervo** [comment]
The needed libssh changes were merged upstream and will be included in the next release in the 0.7 series. So this should work if used with libssh master or on rawhide in a few days. 

I added a configuration variable so that this can be built without public key support if the libssh version is too old at least temporarily while we figure out what requiring libssh > 0.7 will mean on on other OSes.

**Comment by stefwalter** [comment]
@petervo Nice work. The change to libssh looks spot on.

**Comment by stefwalter** [comment]
Fixes #1461

**Comment by petervo** [comment]
Ok this is finally ready for closer review along with tests. I also separated most of the code out of sshtransport as suggested.

To make the tests pass, but had to install libssh from updates-testing since mock builds with updates-testing making the rpms enable key auth. Once 0.7.1 is in updates that won't be needed anymore. I didn't add intergration test yet as that will require a pam module to work properly.

File: configure.ac [hunk 5]
```diff
@@ -432,6 +433,13 @@ AC_MSG_RESULT($enable_strict)
 
 AC_SUBST(REAUTHORIZE_LIBS)
 
+KEY_AUTH=0
+PKG_CHECK_MODULES(LIBSSH_WITH_AGENT_FD, [libssh >= $LIBSSH_KEYAUTH_VERSION],
```
Comment (stefwalter) [inline configure.ac]: I would suggest checking for the required function instead. That's more idiomatic for autotools.

File: src/bridge/bridge.c [hunk 6]
```diff
@@ -478,6 +481,13 @@ run_bridge (const gchar *interactive)
       super = cockpit_portal_new_superuser (transport);
     }
 
+  ssh_auth_sock = g_getenv ("SSH_AUTH_SOCK");
+  if (ssh_auth_sock != NULL && ssh_auth_sock[0] != '\0')
+    {
+      auth_address = g_unix_socket_address_new (ssh_auth_sock);
+      cockpit_channel_internal_address ("ssh-agent", auth_address);
```
Comment (stefwalter) [inline src/bridge/bridge.c]: I think you can immediately unref auth_address.

File: src/ws/cockpitsshagent.c [hunk 7]
```diff
@@ -0,0 +1,314 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2013 Red Hat, Inc.
```
Comment (stefwalter) [inline src/ws/cockpitsshagent.c]: Year.

File: src/ws/cockpitsshagent.c [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+  int status;
+
+  gboolean open;
+  gboolean transport_closed;
+  gboolean pipe_closed;
+  gchar *logname;
+};
+
+G_DEFINE_BOXED_TYPE (CockpitSshAgent, cockpit_ssh_agent, cockpit_ssh_agent_ref, cockpit_ssh_agent_unref);
```
Comment (stefwalter) [inline src/ws/cockpitsshagent.c]: Why not use a full GObject? Too ugly? Or another reason?

File: src/ws/cockpitsshagent.c [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+on_agent_pipe_read (CockpitPipe *pipe,
+                    GByteArray *data,
+                    gboolean end_of_data,
+                    gpointer user_data)
+{
+  CockpitSshAgent *agent = user_data;
+  GBytes *message = NULL;
+  if (!agent->transport_closed)
+    {
+      g_byte_array_ref (data);
```
Comment (stefwalter) [inline src/ws/cockpitsshagent.c]: Using cockpit_pipe_consume() would be more efficient here.

File: src/ws/cockpitsshagent.c [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+  agent->pipe = NULL;
+
+  g_free (agent->channel_id);
+  agent->channel_id = NULL;
+}
+
+int
+cockpit_ssh_agent_get_fd (CockpitSshAgent *agent)
+{
+  return agent->ssh_fd;
```
Comment (stefwalter) [inline src/ws/cockpitsshagent.c]: Where is ssh_fd closed? Does libssh close it? If so, we probably want to change how this function works. Such as cockpit_ssh_agent_claim_fd() and it shuold clear the ssh_fd and have checks against calling it twice, etc.

File: src/ws/cockpitsshagent.h [hunk 11]
```diff
@@ -0,0 +1,49 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2013 Red Hat, Inc.
```
Comment (stefwalter) [inline src/ws/cockpitsshagent.h]: Year.

**Comment by petervo** [comment]
Hubbot seems to be stuck, but I think i addressed all the points in your review.

**Comment by petervo** [comment]
Fixed a couple more bugs, tests are passing now.

File: src/ws/cockpitsshtransport.c [hunk 12]
```diff
@@ -1249,6 +1294,15 @@ cockpit_ssh_transport_constructed (GObject *object)
   g_warn_if_fail (ssh_options_set (self->data->session, SSH_OPTIONS_USER,
                                    cockpit_creds_get_user (self->data->creds)) == 0);
 
+#ifdef HAVE_SSH_SET_AGENT_SOCKET
+  if (self->agent)
+    {
+      int agent_fd = agent_fd = cockpit_ssh_agent_claim_fd (self->agent);
```
Comment (stefwalter) [inline src/ws/cockpitsshtransport.c]: The double assignment here seems strange.

Comment (petervo) [inline src/ws/cockpitsshtransport.c]: fixed

**Comment by stefwalter** [comment]
Ooops @petervo. This breaks when I have an RSA1 style key in my home directory. libssh as used in cockpit-ws should never look in my home directory.

It fails with 'make check'

You should be able to reproduce by running ```ssh-keygen -t rsa1``` in your session.

**Comment by petervo** [comment]
@stefwalter see #2477

**Pull request closed without merging**